crates/rpc/ipc/src/server/connection.rs

```diff
@@ -178,17 +180,56 @@ where
                 }
             }
 
-            // read from the stream
-            match ready!(this.conn.as_mut().poll_next(cx)) {
-                Some(Ok(item)) => {
-                    let call = this.service.call(item);
-                    this.pending_calls.push(call);
+            'inner: loop {
+                let mut drained = false;
+                // drain all calls that are ready and put them in the output item queue
+                if !this.pending_calls.is_empty() {
+                    if let Poll::Ready(Some(res)) = this.pending_calls.as_mut().poll_next(cx) {
+                        let item = match res {
+                            Ok(Some(resp)) => resp,
+                            Ok(None) => continue 'inner,
+                            Err(err) => err.into().to_string(),
+                        };
+                        this.items.push_back(item);
+                        continue 'outer
+                    } else {
+                        drained = true;
+                    }
                 }
-                Some(Err(err)) => {
-                    tracing::warn!("IPC request failed: {:?}", err);
-                    return Poll::Ready(())
+
+                // read from the stream
+                match this.conn.as_mut().poll_next(cx) {
+                    Poll::Ready(res) => match res {
+                        Some(Ok(item)) => {
+                            let mut call = this.service.call(item);
+                            match call.poll_unpin(cx) {
+                                Poll::Ready(res) => {
+                                    let item = match res {
+                                        Ok(Some(resp)) => resp,
+                                        Ok(None) => continue 'inner,
```

Comment:
under which circumstances this can return `Ok(None)`?

Reply:
for subscriptions responses which are routed differently